rfd/0078-trait-transforms.md

```diff
@@ -0,0 +1,762 @@
+  # existing traits, or add new ones.
+  #
+  # All incoming traits are available for use in the expression by
+  # `external.<trait_name>` or `external["<trait_name>"]`.
+  #
+  # Various helper functions are provided, which will be explained later in the
+  # RFD.
+  override:
+    # groups will override the existing groups trait by appending the "dbs"
+    # group if the user's current groups includes "splunk".
+    groups: >
+      match(contains(external.groups, "splunk"),
+        option(true, flatten(external.groups, "dbs")),
+        option(false, external.groups))
+
+    # logins will be a new trait added to the cert.
+    logins: >
+      flatten(
+        "ubuntu",
+
+        regexp.replace(external.username, "-", "_"),
+
+        match(external.email,
+          option("nic@goteleport.com", "root")),
+
+        match(regexp.replace(external.email, "^.*@goteleport.com$", "teleporter"),
+          option("teleporter", email.local(external.email)),
+          default_option("contractor")))
+```
+
+### Predicate Helper Functions
+
+A set of helper functions to be used in the predicate expressions will be
+included.
+Part of the reasoning for using the
+[predicate language](https://github.com/vulcand/predicate)
+is so that it will be extensible.
+As the need arises, we can always add new helper function.
+
+The predicate parser is based on the `Go` language parser, so `Go` keywords
+which may be better names (such as `select`, `default`) must unfortunately be
+avoided.
+
+#### `list(...items)`
+
+`list()` returns a list of all of its arguments.
+All items must be strings.
+
+#### `flatten(...items)`
+
+`flatten()` returns a flattened and deduplicated list of all of its arguments.
+`flatten("a", list("b", "c"), list(), list("a"))`
+will return
+`["a", "b", "c"]`
+
+All items must be strings or lists of strings.
+
+#### `contains(list, value)`
+
+`contains(list, value)` returns `true` if any item in the list is an exact match
+for `value`, else it returns false. `contains(list("a", "b"), "b")` returns
+true.
+
+#### `match(value, ...options)`
+
+`match(value, ...options)` returns the value of the first `option` which is a
+match for `value`.
+
+`match("c", option("a", "foo"), option("b", "bar"), option("c", "baz"))` returns
+`"baz"`
+
+`match(true, option(false, list("a", "b")), option(true, list("c", "d")))`
+returns `["c", "d"]`
+
+`default_option(value)` will match for any input value.
+
+`match("xyz", option("abc", "go"), default_option("stop"))` returns `"stop"`.
+
+We could easily add `regexp_option` which would match based on a regular
+expression, and/or `match_all` which would return the concatenation of all
+matching options.
+
+#### `email.local(email)`
+
+Has the same meaning it currently does in
+[role templates](https://goteleport.com/docs/access-controls/guides/role-templates/#interpolation-rules).
+
+Returns the local part of an email field.
+
+#### `regexp.replace(variable, expression, replacement)`
```

Comment:
@nklaassen Do we need regexp.replace for now? Can we introduce a simple string replace and then evolve it? @nklaassen I will explain the rationale on the follow up call

Reply:
We can certainly go ahead without it for now, I included it because we support it for role templates currently and it seems to be a useful feature.

@klizhentas would you like me to schedule another call?